archey/colors.py

```diff
@@ -12,33 +12,20 @@
 ANSI_ECMA_REGEXP = re.compile(r"\x1b\[\d+(?:(?:;\d+)+)?m")
 
 
-class Colors(Enum):
+class Style:
     """
-    ANSI terminal colors enumeration.
+    Style base-class supporting terminal escape sequences for bold, colour, etc.
     Supports an arbitrary number of display attributes.
-
-    See <https://web.archive.org/web/20200627145120/http://www.termsys.demon.co.uk/vtansi.htm>
-      or <https://en.wikipedia.org/wiki/ANSI_escape_code>.
     """
 
-    CLEAR = (0,)
-    RED_NORMAL = (0, 31)
-    RED_BRIGHT = (1, 31)
-    GREEN_NORMAL = (0, 32)
-    GREEN_BRIGHT = (1, 32)
-    YELLOW_NORMAL = (0, 33)
-    YELLOW_BRIGHT = (1, 33)
-    BLUE_NORMAL = (0, 34)
-    BLUE_BRIGHT = (1, 34)
-    MAGENTA_NORMAL = (0, 35)
-    MAGENTA_BRIGHT = (1, 35)
-    CYAN_NORMAL = (0, 36)
-    CYAN_BRIGHT = (1, 36)
-    WHITE_NORMAL = (0, 37)
-    WHITE_BRIGHT = (1, 37)
-
     def __str__(self):
-        return self.escape_code_from_attrs(";".join(map(str, self.value)))
+        if self.should_color_output():
+            try:
+                return self.escape_code_from_attrs(";".join(map(str, self.value)))  # type: ignore[attr-defined] # pylint: disable=line-too-long
+            except AttributeError:
+                # i.e. someone instantiated `Style` directly...!
+                return ""
```

Comment:
Hey @ingrinder !
Many thanks for the hard work, it's a very clean patch.
My two cents below :

1. I wonder when/why this case occurs ? Do we _really_ need to catch this exception ? 🙄
2. I think we should directly amend @twang817's first commit here, at least for these reasons : it would prevent a huge revision introducing 21+KB additions (and another one removing them just after) from reaching project history (as well as your 153da7f90830a633a5b596774e175f65dc52511a revert could directly "fix up" the first one here). I guess a first commit with your `Style` implementation and another one adding the WWDC 2020 logo afterwards would be optimal. What do you think ?

Thanks again ! Bye 👋

Reply:
I suppose it's not really necessary to catch, but if someone somewhere makes a `Style` and then tries `str` on it, it will raise `AttributeError` unless its value has been explicitly set (which should be ideally from a subclass).
I thought about giving it an `__init__` which sets `value`, but this also causes an `AttributeError` using it as a mixin with an `Enum` class since `value` is immutable in that case.

I suppose we could remove this, probably should be making an instance of `Style` in the first place, although maybe we could write an `__init__(self, *value)` which either checks `isinstance(self, Enum)` or catches an `AttributeError` before setting `value` (EAFP), then it's required to be defined if an instance is made:

```python
    def __init__(self, *value):
        try:
            self.value = value
        except AttributeError:
            # can't set value (probably an `Enum`)
            pass

    def __str__(self):
        if self.should_color_output():
            return self.escape_code_from_attrs(";".join(map(str, self.value)))
        return ""
```

Though admittedly even that feels a little bit hacky :frowning_face:

As for 2, I was trying to avoid force-pushing onto the fork :smile: but let me know which approach you think is best and I'll do an interactive rebase to squash it all into two commits as you suggest (I wasn't sure if you just wanted to squash and merge in one!). :wave:

Reply:
1. Wow, what a mess 👀
    My opinion is that we shouldn't catch it in the first place, if one day someone uses `Style` directly and then calls `str` it will fail and they'll need to fix their code anyway.
    On the other hand, if one day we drop support for "old" Python releases, we could update this class to a more appropriate object (if there is).
3. I'm okay for a squash in one or two commits (preference for two as I suggested, this way we "uncouple" your `Colors` rework/extension from the WWDC logo actual addition) 😉

Many thanks again, cheers 👋